enterprise/views.py

```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
         into an actual EnterpriseCustomer, a set of details about the course, and a list
         of the available course modes for that course run.
         """
+        enterprise_catalog_uuid = request.GET.get('catalog')
         try:
             enterprise_customer = get_enterprise_customer_or_404(enterprise_uuid)
-            course, course_run = \
-                CourseCatalogApiServiceClient(enterprise_customer.site).get_course_and_course_run(course_run_id)
         except ImproperlyConfigured:
             raise Http404
 
+        enterprise_catalog = None
+        if enterprise_catalog_uuid:
+            try:
+                enterprise_catalog = enterprise_customer.enterprise_customer_catalogs.get(
+                    uuid=enterprise_catalog_uuid
+                )
+            except EnterpriseCustomerCatalog.DoesNotExist:
+                add_generic_info_message_for_error(request)
+
+        if enterprise_catalog:
+            course, course_run = enterprise_catalog.get_course_and_course_run(course_run_id)
+        else:
+            try:
+                course, course_run = CourseCatalogApiServiceClient(
+                    enterprise_customer.site
+                ).get_course_and_course_run(course_run_id)
+            except ImproperlyConfigured:
+                raise Http404
+
         if not course or not course_run:
-            LOGGER.warning('Failed to fetch course "{course}" or course run "{course_run}" details'.format(
-                course=course, course_run=course_run
-            ))
+            LOGGER.warning(
+                'Failed to fetch course "{course}" or course run "{course_run}" details for '
+                'enterprise "{enterprise_uuid}"'.format(
+                    course=course, course_run=course_run, enterprise_uuid=enterprise_customer.uuid
+                )
+            )
             raise Http404
 
+        course_modes = []
+        if enterprise_catalog_uuid and not enterprise_catalog:
+            return enterprise_customer, course, course_run, course_modes
+
         modes = EnrollmentApiClient().get_course_modes(course_run_id)
         if not modes:
             LOGGER.warning('Unable to get course modes for course run id {course_run_id}.'.format(
                 course_run_id=course_run_id
             ))
             raise Http404
 
-        course_modes = []
+        if enterprise_catalog:
+            # filter and order course modes according to the enterprise catalog
+            # modes = filter(
+            #     lambda course_mode: course_mode['slug'] in enterprise_catalog.enabled_course_modes, modes
+            # )
+            modes = [mode for mode in modes if mode['slug'] in enterprise_catalog.enabled_course_modes]
+            modes.sort(key=lambda course_mode: enterprise_catalog.enabled_course_modes.index(course_mode['slug']))
+            if not modes:
+                add_generic_info_message_for_error(request)
```

Comment:
Move 548-563 to a separate function: `get_available_course_modes(course_run_id, enterprise_customer_catalog=enterprise_catalog)`